### DISCUSSION TITLE
Where is the best place to `createMachine` in React?

### DISCUSSION THREAD

**Discussion by samzmann** [statelyai/xstate#2206] [Q&A]
I have a few machines in my react app. I instantiate them as global variables with `createMachine`, as recommended [here](https://xstate.js.org/docs/recipes/react.html).
Some of these machines are only used in obscure flows that the user rarely reaches, so I'm wondering if there is a way to "defer" instantiating the machine until I mount the component that actually needs it.

What I tried is to change the variable to a function that returns a state machine, eg:
```JavaScript
// change this...
export const toggleMachine = createMachine({ ... })

// ...to this
export const generateToggleMachine = () => {
  console.log('generating machine')
  return createMachine({ ... })
}
```

and then call it like this
```JavaScript
const [machine, send] = useMachine(generateToggleMachine())

// or 

const toggleMachine = useRef(generateToggleMachine()).current
const [machine, send] = useMachine(toggleMachine)
```

These two options seem to work: the machine functions as expected and I cannot see any issues, but `"generating machine"` is logged on every render which I did not expect...

So my questions are:
- is having many state machines globally an issue, and do I really need to do anything about it?
- if so, what is the best practice way to instantiate a state machine on component mount?
- is my current solution safe, despite seeing the `"generating machine"` log on each render?

**Comment by VanTanev** [comment]
`useMachine()` takes  machine or a function that lazily creates a machine, so you can do:

```typescript
const [state, send] = useMachine(() => createMachineFactory())
```

I think the general best practice is to use this:
https://github.com/vantanev/xstate-helpers#createreactcontexthelpers ;)
